text/2021-09-09-physical-machine-auth.md

```diff
@@ -0,0 +1,189 @@
+
+#### User Cases
+
+Here is a description of three different usage scenarios of user cases.
+
+#### Case1: Automatically generate certificates
+
+Prerequisites:
+
+1. User deployed `Chaos Mesh` in Kubernetes cluster with security mode
+1. The node executing the `chaosctl` command can ssh to
+the target physical machine
+1. The node executing the `chaosctl` command can access the Kubernetes cluster
+
+Steps:
+
+1. User prepares the physical machine using `chaosctl`, the command might be
+`chaosctl physical-machine init --server=127.0.0.1 --port=2333`, in this step,
+`chaosctl` generates the certificates on the physical machine side and copies
+all the required certificates to the target physical machine, then create the
+`PhysicalMachine` CR in Kubernetes cluster (BTW, `chaosctl pm` can be used
+instead of `chaosctl physical-machine`)
+1. User starts chaosd service on the physical machine
+1. User creates a physical machine experiment on the dashboard
+1. Chaos-controller-manager establishes an mTLS connection when requesting
+the chaosd service on the physical machine
+
+#### Case2: Manually generate certificates
+
+Prerequisites:
+
+1. User deployed `Chaos Mesh` in Kubernetes cluster with security mode
+1. The node executing the `chaosctl physical-machine create` command can
+access the Kubernetes cluster
+
+Steps:
+
+1. User copies the CA certificate from the kubernetes cluster
+to the physical machine
+1. User uses `chaosctl` to generate the certificates on the physical machine,
+the command might be `chaosctl physical-machine generate`
+1. User uses `chaosctl` to create `PhysicalMachine` resource in Kubernetes
+cluster, the command might be
+`chaosctl physical-machine create --server=127.0.0.1 --port=2333`
+1. User starts chaosd service on the physical machine
+1. User creates a physical machine experiment on the dashboard
+1. Chaos-controller-manager establishes an mTLS connection when requesting
+the chaosd service on the physical machine
+
+#### Case3: Without mTLS authentication (Not recommended)
+
+Prerequisites:
+
+1. User deployed `Chaos Mesh` in Kubernetes cluster without security mode
+
+Steps:
+
+1. User uses `chaosctl` to create `PhysicalMachine` resource in Kubernetes
+cluster, the command might be
+`chaosctl physical-machine create --server=127.0.0.1 --port=2333 --protocol=http`
+1. User starts chaosd service on the physical machine
+1. User creates a physical machine experiment on the dashboard
+1. Chaos-controller-manager will use HTTP to request the chaosd service
+
+## Drawbacks
+
+Because `PhysicalMachine` is a namespaced scope resource, users can create
+the same physical machine information on different namespaces, which may
+cause duplicate injection problems and  it requires the chaosd service
+API to support idempotency.
+
+## Alternatives
+
+NA
+
+## Unresolved questions
+
+### Not support a role to access multiple specified `PhysicalMachines`
```

Comment:
how about creating the related CRD of `PhysicalMachine` under the same namespace?

Reply:
The problem here is that this solution still relies on the Kubernetes RBAC implementation, so the definition of roles is still only two resources, `Role` and `ClusterRole`. 
1. Using `Role`, the user only has the access of one namespace, and with global considerations, can only experiment with pods in that namespace, this may not be the result we want.
2. Using `ClusterRole`, the user has cross-namespace access, so it is not possible to restrict. 

In short, I think this is due to the mismatch between the Kubernetes RBAC mechanism and this requirement. Kubernetes only controls access at the namespace level, and it is not suitable for more granular control of resource access. And most importantly, we don't have a clear understanding of the actual user's access control needs for now, so it's hard to determine which solution is more appropriate.